### PR TITLE
Small cleanup to gradient index and hist.

### PR DESCRIPTION
**Pull request by trivialfis**
Extracted from https://github.com/dmlc/xgboost/pull/7659

* Code comments.
* Const accessor to index.
* Remove/rename some variables in the `Index` class.
* Simplify the `MemStackAllocator`.

### CODE REVIEW

**Comment by trivialfis** [comment]
cc @ShvetsKS

**Comment by trivialfis** [comment]
@hcho3 @RAMitchell Please take a look into this PR when you are available.  Full impl for hist cat support is in https://github.com/dmlc/xgboost/pull/7659 .

File: src/common/threading_utils.h [hunk 1]
```diff
@@ -246,6 +246,43 @@ inline int32_t OmpGetNumThreads(int32_t n_threads) {
...
+  ~MemStackAllocator() {
+    if (required_size_ > MaxStackSize) {
+      free(ptr_);
+    }
+  }
+  T& operator[](size_t i) { return ptr_[i]; }
+  T const& operator[](size_t i) const { return ptr_[i]; }
+
+  // FIXME(jiamingy): Remove this once we merge partitioner cleanup for hist.
+  auto Get() { return ptr_; }
```
Comment (hcho3) [inline src/common/threading_utils.h:279]: Can we remove this now? `GHistIndexMatrix` is no longer using `Get()` method.

**Review by hcho3 — APPROVED** [review]

File: src/common/threading_utils.h (hunk 1 above)
Comment (trivialfis) [reply src/common/threading_utils.h:279]: @hcho3 It's still used by the hist updater.

**Pull request merged**